### PR TITLE
feat: Variants Assembly

### PR DESCRIPTION
**Pull request by kendallstrautman**
## 🔗 Relevant links

- [Preview link](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/) 🔎
- [Asana task](https://app.asana.com/0/1204333057896641/board) 🎟️

## 🗒️ What

This Assembly PR contains all the frontend work to support tutorial variants. See the[ asana project](https://app.asana.com/0/1204333057896641/board) and [RFC](https://docs.google.com/document/d/1bDQCGDgjJouBAOfc8K5zkjeSjOFvg8gC7wXPeBasf8E/edit#) for full context. 

### Assembles these PRs

- https://github.com/hashicorp/dev-portal/pull/1912
- https://github.com/hashicorp/dev-portal/pull/1928
- https://github.com/hashicorp/dev-portal/pull/1931
- https://github.com/hashicorp/dev-portal/pull/1938
- https://github.com/hashicorp/dev-portal/pull/1937
- https://github.com/hashicorp/dev-portal/pull/1946
- https://github.com/hashicorp/dev-portal/pull/1952
- https://github.com/hashicorp/dev-portal/pull/1953
- https://github.com/hashicorp/dev-portal/pull/1958
- https://github.com/hashicorp/dev-portal/pull/1954
- https://github.com/hashicorp/dev-portal/pull/1959
- https://github.com/hashicorp/dev-portal/pull/1960

### Related but unmerged PR 
- https://github.com/hashicorp/dev-portal/pull/1961

### Variant enabled paths

- [/waypoint/tutorials/get-started-nomad-cloud/get-started-nomad-cloud](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/waypoint/tutorials/get-started-nomad-cloud/get-started-nomad-cloud)
- [/vault/tutorials/getting-started/getting-started-install](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/vault/tutorials/getting-started/getting-started-install)
- [/consul/tutorials/get-started-kubernetes/kubernetes-gs-deploy](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/consul/tutorials/get-started-kubernetes/kubernetes-gs-deploy) (has actual tab content updated to use variants)
- [/consul/tutorials/kubernetes/kubernetes-api-gateway](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway) (actual tab content updated here as well)

### Steps
- Visit the paths above. You should see variants UI in the tutorial metadata and in the sidecar.
- Select a variant option
- Navigate away and back, validate that your option selection persists, can check your cookies for `variants` to see the data.
- Visit a tutorial with the same options, validate that the state persists between both when an option is selected: [/consul/tutorials/get-started-kubernetes/kubernetes-gs-deploy](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/consul/tutorials/get-started-kubernetes/kubernetes-gs-deploy) & [/consul/tutorials/kubernetes/kubernetes-api-gateway](https://dev-portal-git-ksvariants-assembly-hashicorp.vercel.app/consul/tutorials/kubernetes/kubernetes-api-gateway)
- Validate that tutorials without variants are unaffected

## 💭 Anything else?

There was a transient issue noted during QA related to stale variant state when navigating via next/link in the sidebar quickly. [A ticket is here for that](https://app.asana.com/0/1204333057896641/1204685816482754), it remains unresolved but difficult to reproduce.

### CODE REVIEW

File: src/components/sidebar/components/sidebar-nav-menu-item/index.tsx [hunk 1]
```diff
@@ -100,9 +100,24 @@ const SidebarNavLinkItem = ({ item }: SidebarNavLinkItemProps) => {
 
 	// Determine the trailing icon to use, if any
 	const trailingIcon = isExternal ? <IconExternalLink16 /> : item.trailingIcon
-
-	// Conditionally determining props for the <a>
 	const ariaCurrent = !isExternal && item.isActive ? 'page' : undefined
+	const [isMounted, setIsMounted] = useState(false)
+
+	/**
+	 * Note from on this useEffect:
```
Comment (kendallstrautman) [inline src/components/sidebar/components/sidebar-nav-menu-item/index.tsx:107]:
```suggestion
	 * Note on this useEffect:
```

File: src/lib/learn-client/api/tutorial/formatting.ts [hunk 2]
```diff
@@ -48,6 +56,13 @@ export function formatTutorialData(
 		featured_collections,
 		default_collection_id
 	)
+	// We only accept a single variant currently, so we take the first array item
+	// The learn-api-content-sync should prevent multiple variants from syncing into
+	// the database, this is an extra protection
+	// @TODO once learn api variants updates are released, remove this check
+	// right now the prod api doesn't return tutorials with an emtpy variants array
+	const formattedVariant =
+		variants && variants.length > 0 ? formatVariant(variants[0]) : undefined
```
Comment (kendallstrautman) [inline src/lib/learn-client/api/tutorial/formatting.ts:65]:
```suggestion
	const formattedVariant = variants?.length > 0 ? formatVariant(variants[0]) : undefined
```

File: src/views/tutorial-view/components/variant-dropdown-disclosure/index.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	getVariantParam,
+	getVariantPath,
+	handleVariantCookie,
+	sortVariantOptions,
+	TutorialVariantOption,
+} from 'views/tutorial-view/utils/variants'
+import { VariantDropdownDisclosureProps } from './types'
+import s from './variant-dropdown-disclosure.module.css'
+
+/** @TODO add onclick for cookie setting */
```
Comment (kendallstrautman) [inline src/views/tutorial-view/components/variant-dropdown-disclosure/index.tsx:17]:
```suggestion
```

File: src/views/tutorial-view/components/variant-dropdown-disclosure/types.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { TutorialVariant } from 'views/tutorial-view/utils/variants/types'
+
+export interface VariantDropdownDisclosureProps {
+	variant: TutorialVariant
+	classNames?: {
+		dropdownRoot: string
+		dropdownActivator: string
+	}
```
Comment (kendallstrautman) [inline src/views/tutorial-view/components/variant-dropdown-disclosure/types.ts:8]:
```suggestion
```
I don't think this is used anymore

File: src/views/tutorial-view/utils/variants/index.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	if (tutorialVariant) {
+		// find the default variant base on the order number
+		const defaultVariantOption = tutorialVariant.options.find(
+			(option: ClientTutorialVariantOption) => option.displayOrder === 1
+		)
+		let activeOption = defaultVariantOption
+
+		// if the variant slug is passed via query param, use that, otherwise pass the default
+		if (variantSlug) {
+			// slugs in the query params are formatted like ?variants=slug:optionSlug
```
Comment (kendallstrautman) [inline src/views/tutorial-view/utils/variants/index.ts:82]:
```suggestion
			// slugs in the query params are formatted like `slug:optionSlug`
```

File: src/views/tutorial-view/utils/variants/types.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import {
+	TutorialVariant as ClientTutorialVariant,
+	TutorialVariantOption as ClientTutorialVariantOption,
+} from 'lib/learn-client/types'
+
+// @TODO rename to TutorialViewVariant? Wait until rebase to avoid conflicts
```
Comment (kendallstrautman) [inline src/views/tutorial-view/utils/variants/types.ts:6]:
```suggestion
```

**Review by BrandonRomano — APPROVED** [review]
🎉

**Pull request merged**